File: solutions/589_N-aryTreePreorderTraversal.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""
class Solution(object):
    def preorder(self, root):
        """
        :type root: Node
        :rtype: List[int]
        """
        result = []
        self.dfs_rec(root, result)
        return result
    
    def dfs_rec(self, root, result):
        if not root:
            return
        
        result.append(root.val)
        for child in root.children:
            self.dfs_rec(child,result)

File: solutions/589_N-aryTreePreorderTraversal.js
/**
 * // Definition for a Node.
 * function Node(val,children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */
/**
 * @param {Node} root
 * @return {number[]}
 */
var preorder = function(root) {
    var result = [];
    
    function helper(node){
        if(node){
            result.push(node.val)
            for(let i = 0; i<node.children.length;i++){
                helper(node.children[i]);
            }
        }
    }
    helper(root);
    return result
};